embedded-tests/src/test/java/org/apache/druid/testing/embedded/query/QueryErrorTest.java

```diff
@@ -0,0 +1,282 @@
+import org.junit.jupiter.params.provider.FieldSource;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.druid.testing.embedded.query.ServerManagerForQueryErrorTest.QUERY_CAPACITY_EXCEEDED_TEST_CONTEXT_KEY;
+import static org.apache.druid.testing.embedded.query.ServerManagerForQueryErrorTest.QUERY_FAILURE_TEST_CONTEXT_KEY;
+import static org.apache.druid.testing.embedded.query.ServerManagerForQueryErrorTest.QUERY_TIMEOUT_TEST_CONTEXT_KEY;
+import static org.apache.druid.testing.embedded.query.ServerManagerForQueryErrorTest.QUERY_UNSUPPORTED_TEST_CONTEXT_KEY;
+import static org.apache.druid.testing.embedded.query.ServerManagerForQueryErrorTest.RESOURCE_LIMIT_EXCEEDED_TEST_CONTEXT_KEY;
+
+/**
+ * This class tests various query failures.
+ * <p>
+ * - SQL planning failures. Both {@link org.apache.calcite.sql.parser.SqlParseException}
+ * and {@link org.apache.calcite.tools.ValidationException} are tested using SQLs that must fail.
+ * - Various query errors from historicals. These tests use {@link ServerManagerForQueryErrorTest} to make
+ * the query to always throw an exception.
+ */
+public class QueryErrorTest extends QueryTestBase
+{
+  //  TODO: introduce onAnyRouter(...) and use it; add TLS tests in the follow-up patches
+  protected static List<Boolean> SHOULD_USE_SQL_ENGINE = List.of(true, false);
+
+  @Override
+  protected EmbeddedDruidCluster createCluster()
+  {
+    overlord.addProperty("druid.manager.segments.pollDuration", "PT0.1s");
+    indexer.setServerMemory(600_000_000)
+           .addProperty("druid.worker.capacity", "4")
+           .addProperty("druid.processing.numThreads", "2")
+           .addProperty("druid.segment.handoff.pollDuration", "PT0.1s");
+
+    return EmbeddedDruidCluster.withEmbeddedDerbyAndZookeeper()
+                               .useLatchableEmitter()
+                               .addServer(overlord)
+                               .addServer(coordinator)
+                               .addServer(broker)
+                               .addServer(router)
+                               .addServer(indexer)
+                               .addServer(historical)
+                               .addExtension(ServerManagerForQueryErrorTestModule.class);
+  }
+
+  @Override
+  protected void beforeAll()
+  {
+    var ingestionStatus = cluster.callApi().onAnyBroker(
+        b -> b.submitSqlTask(
+            new ClientSqlQuery(
+                "REPLACE INTO t\n"
+                + "OVERWRITE ALL\n"
+                + "SELECT CURRENT_TIMESTAMP AS __time, 1 AS d\n"
+                + "PARTITIONED BY ALL",
+                null,
+                false,
+                false,
+                false,
+                Map.of(),
+                List.of()
+            )
+        )
+    );
+    cluster.callApi().waitForTaskToSucceed(ingestionStatus.getTaskId(), overlord);
+    try {
+      Thread.sleep(1000L);
```

Comment:
Why is this sleep needed?

Reply:
Ah, a comment has been lost for some reason. Actually, it is a commit message :D 

This sleep is needed to prolong the time distance between the ingestion task **end** and test method **start**. 
Even though the method on line 98 is named `waitForTaskToSucceed`, in fact some tests are started before the task succeeds and they fail with 'data source is not available'. 
I may just assume, that it is ended on overlord, but broker/router still don't get updated metadata. 
From top of your head, is there any check/call to ensure that data source exists?